crates/ruff_linter/src/checkers/ast/mod.rs

```diff
@@ -2348,13 +2355,17 @@ impl<'a> Checker<'a> {
                     }
                     self.parsed_type_annotation = None;
                 } else {
-                    if self.enabled(Rule::ForwardAnnotationSyntaxError) {
-                        self.diagnostics.push(Diagnostic::new(
-                            pyflakes::rules::ForwardAnnotationSyntaxError {
-                                body: string_expr.value.to_string(),
-                            },
-                            string_expr.range(),
-                        ));
+                    self.semantic.restore(snapshot);
+
+                    if !self.semantic.in_no_type_check() {
+                        if self.enabled(Rule::ForwardAnnotationSyntaxError) {
+                            self.diagnostics.push(Diagnostic::new(
+                                pyflakes::rules::ForwardAnnotationSyntaxError {
+                                    body: string_expr.value.to_string(),
+                                },
+                                string_expr.range(),
+                            ));
```

Comment:
Should we instead add a `push_type_diagnostic` method to `Checker` so that we don't need to repeat the `!self.semantic.in_no_type_check` check. This would also be closer to how this works in Red Knot.

Reply:
That's fair. But having the method would prepare us to add `type: ignore` support because it would provide a centralized place to where such a check could be added (vs changing all call sites)